### ISSUE TITLE
Field name with square brackets

### ISSUE THREAD

**Issue by pnevyk** [murar8/axum_typed_multipart#10]
It seems that it's a common practice to use trailing square brackets (e.g., `names[]=Foo`) when one wants to send multiple values with the same field name. For examples, here is the PHP manual for [uploading multiple files](https://www.php.net/manual/en/features.file-upload.multiple.php) and I found the same recommendations during internet search ([1](https://www.darklaunch.com/curl-form-upload-multiple-files.html), [2](https://stackoverflow.com/questions/21124450/how-to-use-curl-multipart-form-data-to-post-array-field-from-command-line)).

This would mean that for the following program:

```rust
use axum::{http::StatusCode, routing::post, Router};
use axum_typed_multipart::{TryFromMultipart, TypedMultipart};

#[derive(TryFromMultipart)]
struct RequestData {
    names: Vec<String>,
}

async fn handler(TypedMultipart(RequestData { names }): TypedMultipart<RequestData>) -> StatusCode {
    println!("names = {:?}", names);
    StatusCode::OK
}

#[tokio::main]
async fn main() {
    // build our application with a single route
    let app = Router::new().route("/", post(handler));

    // run it with hyper on localhost:3000
    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}
```

the following request:

```
curl -X POST --form "names[]=Foo" --form "names[]=Bar" http://localhost:3000
```

prints `names = ["Foo", "Bar"]`.

[Rocket](https://rocket.rs/) and its [FromForm](https://docs.rs/rocket/0.5.0-rc.3/rocket/form/trait.FromForm.html) derive trait supports this.

I think that supporting just `names[]` is a good start, later it would be nice to support things like `names[0]` (explicit ordering) or even `names[foo]` -> HashMap.

If this is desired feature, I could try to send a PR. In that case and advice and pointers into the codebase would be appreciated.

**Comment by murar8** [comment]
Hi, I took a look at the Rocket docs and it looks like a really cool idea but it would be quite a bit of work so I'm not sure I will be able to work on it at the moment.

**Comment by pnevyk** [comment]
Sure, take your time, it's not a blocker for me. Thanks!

The advanced features that Rocket supports are nice, but I wonder how much they are useful in practice.

**Comment by murar8** [comment]
Yeah that's a good point, I guess we could start with the `names[]` and `names[0]` convention for arrays, that would probably cover 99% of use cases

**Comment by gaurav-bagga** [comment]
Nested structs too please.
Sample use case, perhaps possible.

```
struct Student {
  name: String,
  avatar: FieldData<Bytes>
}


struct School {
   name: String,
   students: Vec<Student>,
   avatar: FieldData<Bytes>
}
```

**Comment by murar8** [comment]
@pnevyk @gaurav-bagga feel free to take a look at the linked PR, let me know if the feature seems complete or if it's missing anything to make it nice to use.

**Comment by pnevyk** [comment]
@murar8 Unless I am doing something wrong, I believe that the implementation doesn't support the following use case:

```rust
#[derive(TryFromMultipart)]
struct CreateTeamRequest {
    members: Vec<String>, // Array of standard types
}

#[tokio::test]
async fn test_multivalue_success() {
    let client = TestClient::new(nested::app());
    let form = Form::new()
        .text("members[]", "Alice")
        .text("members[]", "Bob");
    let res = client.post("/teams").multipart(form).send().await.unwrap();
    assert_eq!(res.status(), StatusCode::CREATED);
}
```

The indexes should be optional – the default behavior is to push each value to the `Vec`. Not sure how to handle mixing explicit and implicit indexes though.

Also, the current implementation doesn't seem to support standard types like `String` and integers in `Vec`. If I use `#[form_data(nested)]` on the field, I get _cannot find type `StringMultipartBuilder` in this scope_ error. If I don't use the attribute, I get invalid request.

Is it possible to support this?

**Comment by murar8** [comment]
@pnevyk those are all great observations, thanks for taking the time. The problem with the way I implemented lists on top level is that arrays have optional indexes which creates ambiguity.

I'm not quite sure what would be best in this case, for example Rocket solves it like this:

```rust
// These form strings...
"numbers[]=1&numbers[]=2&numbers[]=3",
"numbers[a]=1&numbers[b]=2&numbers[c]=3",
"numbers[a]=1&numbers[b]=2&numbers[a]=3",
"numbers[]=1&numbers[b]=2&numbers[c]=3",
"numbers.0=1&numbers.1=2&numbers[c]=3",
"numbers=1&numbers=2&numbers=3",

// ...parse as this struct:
MyForm {
    numbers: vec![1 ,2, 3]
}
```

Let me rework the feature a bit and see if I can make it more consistent with top-level arrays.
